source/Calamari/Kubernetes/Conventions/HelmUpgradeWithKOSConvention.cs

```diff
@@ -0,0 +1,554 @@
+            ct.CancelAfter(timeout);
+            string manifest = null;
+            while (!ct.IsCancellationRequested)
+            {
+                try
+                {
+                    manifest = helmCli.GetManifest(releaseName, revisionNumber);
+                    log.Verbose($"Retrieved manifest for {releaseName}, revision {revisionNumber}.");
+                    break;
+                }
+                catch (CommandLineException)
+                {
+                    log.Verbose("Helm manifest was not ready for retrieval. Retrying in 1s.");
+                    await Task.Delay(TimeSpan.FromSeconds(1), ct.Token);
+                }
+            }
+
+            if (string.IsNullOrWhiteSpace(manifest))
+            {
+                throw new CommandException("Failed to retrieve helm manifest in a timely manner");
+            }
+
+            return manifest;
+        }
+
+        async Task ParseManifestAndMonitorResourceStatuses(RunningDeployment deployment, string manifest, CancellationToken cancellationToken)
+        {
+            var resources = new List<ResourceIdentifier>();
+            using (var reader = new StringReader(manifest))
+            {
+                var yamlStream = new YamlStream();
+                yamlStream.Load(reader);
+
+                foreach (var document in yamlStream.Documents)
+                {
+                    if (!(document.RootNode is YamlMappingNode rootNode))
+                    {
+                        log.Warn("Could not parse manifest, resources will not be added to kubernetes object status");
+                        continue;
+                    }
+
+                    var kind = rootNode.GetChildNode<YamlScalarNode>("kind").Value;
+
+                    var metadataNode = rootNode.GetChildNode<YamlMappingNode>("metadata");
+                    var name = metadataNode.GetChildNode<YamlScalarNode>("name").Value;
+                    var @namespace = metadataNode.GetChildNodeIfExists<YamlScalarNode>("namespace")?.Value;
+
+                    //if the resource doesn't have a namespace set in the manifest set it to the helm namespace.
+                    //This is because namespaced resources that don't have the namespace defined in the manifest will be set in the namespace set in the helm command
+                    //if this is null, we'll fall back on the namespace defined for the kubectl tool (which is the default target namespace)
+                    //we aren't changing the manifest here, just changing where the kubectl looks for our resource.
+                    //We also try and filter out known non-namespaced resources
+                    if (string.IsNullOrWhiteSpace(@namespace) && !KubernetesApiResources.NonNamespacedKinds.Contains(kind))
+                    {
+                        @namespace = deployment.Variables.Get(SpecialVariables.Helm.Namespace)?.Trim();
+                    }
```

Comment:
We have a fixed list of items that are known to not be namespaced, so we don't include the namespace if we _know_ they aren't namespaced.

For all other types, we just fallback to using the helm namespaced variable

Reply:
This sounds related to the work @liam-mackie was doing around dealing with non-namespaced resources. 

It's not _likely_ that customers will run into it here, but it is possible I guess for them to be creating a non-namespaced resource that we don't know about.

I'm wondering if we should fallback to the similar logic we were _going_ to go with these resources and check with the known resource types for their namespaced-ness.

Not something I'd block the PR over, but id like to make sure we are happy with the potential gap

Reply:
What about custom resources that are cluster scoped (not namespaced)?

Reply:
because we use `kubectl`, it will actually ignore it if it knows it's not namespaced.

However, we actually ignore non-namespaced resources completely

This code above mainly just affects the UI and what resources we tell you we are tracking.

This is not a complete solution, but as I said, non-namespaced resources aren't supported anyway, so we need a better solution here.